### PR TITLE
JSUI-2705 Add CategoryFacet values styling (& more!)

### PR DESCRIPTION
**Pull request by ThibodeauJF**
https://coveord.atlassian.net/browse/JSUI-2705

[Deploy](https://dashboard.heroku.com/pipelines/a3535101-5bbf-4a5b-a909-47fcf8c9f149)

### CODE REVIEW

File: sass/CategoryFacet/_CategoryFacetValues.scss [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+  }
+}
+
+.coveo-dynamic-category-facet-collapsed {
+  .coveo-dynamic-category-facet-value:not(.coveo-selected) {
+    display: none;
+  }
+}
+
+.coveo-dynamic-category-facet-value,
```
Comment (olamothe) [inline sass/CategoryFacet/_CategoryFacetValues.scss]: Is there some repetition between these (and, actually this whole file), and the "standard" DynamicFacet values ?

Could we extract some mixins between the two types of facet values that we could leverage in both files ?

File: src/ui/CategoryFacet/CategoryFacetValues/CategoryFacetValueRenderer.ts [hunk 2]
```diff
@@ -1,32 +1,51 @@
...
-    this.dom.on('click', () => this.selectAction());
-    this.toggleSelectedClass();
-    return this.dom.el;
+    this.button.toggleClass('coveo-selected', this.facetValue.isSelected);
+    this.button.toggleClass('coveo-with-space', this.shouldHaveMargin);
+    this.facetValue.isSelected && this.button.setAttribute('disabled', 'true');
+    this.shouldHaveArrow && this.prependArrow();
+
+    this.button.on('click', () => this.selectAction());
+    return $$('li', { dataValue: this.facetValue.value }, this.button).el;
```
Comment (olamothe) [inline src/ui/CategoryFacet/CategoryFacetValues/CategoryFacetValueRenderer.ts]: I'd break this function down a bit just for readability purpose.

eg:
```
this.renderButton()
this.renderFacetLabel()
this.renderFacetCount()
etc..
```

File: unitTests/ui/CategoryFacet/CategoryFacetValues/CategoryFacetValueRendererTest.ts [hunk 3]
```diff
@@ -1,27 +1,141 @@
...
-    // TODO: add XSS tests
     it('should render without errors', () => {
       expect(() => facetValueRenderer.render()).not.toThrow();
     });
+
+    it('should not render XSS in the displayValue', () => {
+      const fakeFacetValueWithXSS = CategoryFacetTestUtils.createFakeFacetValue();
+      fakeFacetValueWithXSS.displayValue = '<script>alert("Hehe goodbye")</script>';
+      initializeComponentWithValue(fakeFacetValueWithXSS);
+      expect(getLabel().textContent).toBe(fakeFacetValueWithXSS.displayValue);
```
Comment (olamothe) [inline unitTests/ui/CategoryFacet/CategoryFacetValues/CategoryFacetValueRendererTest.ts]: A good test would be the check that the node `innerHTML` is "escaped" as text.

File: sass/CategoryFacet/_CategoryFacetValues.scss (hunk 1 above)
Comment (ThibodeauJF) [reply sass/CategoryFacet/_CategoryFacetValues.scss]: A few things but yes, done

File: src/ui/CategoryFacet/CategoryFacetValues/CategoryFacetValues.ts [hunk 4]
```diff
@@ -87,44 +93,50 @@ export class CategoryFacetValues {
...
-      facetValue.children = path[level] ? facetValue.children.filter(child => Utils.arrayEqual(targetPath, child.path)) : [];
-      this.collapseHierarchyAtPathLevel(facetValue.children, path, level + 1);
+      facetValue.children = this.collapseHierarchyAtPathLevel(facetValue.children, path, level + 1);
     });
+
+    return collapsedFacetValues;
+  }
+
+  private collapseHierarchyWithPath(path: string[]) {
+    this.facetValues = this.collapseHierarchyAtPathLevel(this.facetValues, [...path]);
```
Comment (btaillon) [inline src/ui/CategoryFacet/CategoryFacetValues/CategoryFacetValues.ts]: Is cloning the path still relevant? It doesn't seem like `collapseHierarchyAtPathLevel` modifies `path`.

File: src/controllers/DynamicCategoryFacetQueryController.ts [hunk 5]
```diff
@@ -49,7 +49,7 @@ export class DynamicCategoryFacetQueryController {
       field: this.facet.fieldName,
       type: this.facet.facetType,
       currentValues: this.currentValues,
-      numberOfValues: this.numberOfValuesToRequest,
+      numberOfValues: this.facet.values.hasSelectedValue ? 1 : this.numberOfValuesToRequest,
```
[the remark this answers is not in the document]
Comment (ThibodeauJF) [reply src/controllers/DynamicCategoryFacetQueryController.ts]: Once a value is selected you start drilling down it's tree. You don't need to fetch values at the same level anymore.

File: src/ui/CategoryFacet/CategoryFacetValues/CategoryFacetValues.ts (hunk 4 above)
Comment (ThibodeauJF) [reply src/ui/CategoryFacet/CategoryFacetValues/CategoryFacetValues.ts]: You're right!

File: src/ui/CategoryFacet/CategoryFacetValues/CategoryFacetValues.ts [hunk 6]
```diff
@@ -87,44 +93,50 @@ export class CategoryFacetValues {
...
+    const siblingValues = path.length === 1
+      ? this.facetValues
+      : this.getOrCreateFacetValueWithPath(path.slice(0, -1)).children;
 
-    const newFacetValue = this.createFacetValueAtPath(path);
+    siblingValues.push(newFacetValue)
     return newFacetValue;
   }
 
   private collapseHierarchyAtPathLevel(facetValues: CategoryFacetValue[], path: string[], level = 1) {
```
[the remark this answers is not in the document]
Comment (ThibodeauJF) [reply src/ui/CategoryFacet/CategoryFacetValues/CategoryFacetValues.ts]: Level is not confusing if you've played Super Mario :trollface: 
Honestly it's quite clear enough to me and it's already used a number of times the renaming would be a waste of time

I agree for filter thought!

**Pull request merged**